Add option to preserve image aspect ratio

michaelneverwins opened the pull request:
This resolves #177 by adding an option to preserve image aspect ratio by cropping to 2:3 before resize instead of stretching. The new option is presented as a switch to allow or disallow stretching, and this switch is turned on by default to preserve the current behavior.

![preferences](https://github.com/kra-mo/cartridges/assets/73746653/c2b5f933-9d9c-4735-bf86-23616d0073f0)

I tested the new setting by adding the following square (200x200) image with and without stretching disabled:

![square-pepper](https://github.com/kra-mo/cartridges/assets/73746653/c7b38140-b913-45c8-a8a7-5ea887eac2b1)

The result can be seen here:

![comparison](https://github.com/kra-mo/cartridges/assets/73746653/1ee5f6b0-2840-4c4b-b390-ad2b19f3c971)

Obviously a bit of the image is lost when cropping. This can be less than ideal when the original dimensions are wide enough that a significant amount of the image is lost, but stretching a significantly wider-than-2:3 image also has its drawbacks. In either case, using images with an aspect ratio somewhat close to 2:3 is best.

As a more practical example, we can use the official box art for _Darkula_ (from https://locomalito.com/darkula.php), which is 600x850 — _close_ to 2:3 but not exactly — and notably displays a circle.

![comparison-darkula](https://github.com/kra-mo/cartridges/assets/73746653/2a9b59df-4489-4ec2-a0f0-e1b2b9a34f51)

I can tell that the circle is no longer perfectly circular with the default setting. With stretching disabled, the aspect ratio of the circle is preserved, and no valuable image content is lost.

michaelneverwins commented:
Note that I might not be using exactly the right terminology here. The new option does not maintain image aspect ratio in the sense that the 200x200 pepper image remains 1:1, for example; it still ends up as a 2:3 image. Meanwhile, the image in this case is still "stretched" in the sense that it gets enlarged to 200x300 after being cropped to 134x200. The point is that it doesn't get stretched vertically to make the original 200x200 area fill the 200x300 size. In the case of the originally 600x850 _Darkula_ cover, I suppose what I've prevented is more like horizontal squishing than vertical stretching.

The main reason I've attempted to implement this myself (besides fun) was to clarify exactly what I meant in #177, as I think the code I've written does exactly what I want (give or take a pixel because I didn't pay much attention to rounding). But even if the code works, the text I chose for the option in the preferences menu could probably be reworded for clarity.

kra-mo commented:
This is not touching the right part of code, nor would I want to implement such a thing. This is a decision that could only be made game cover by game cover. I think our current solution of only stretching covers below a certain threshold works. Please wait for a response before opening a PR in the future.

michaelneverwins commented:
It seems to me that this was touching the right part of the code for what I intended to do, considering the `CoverManager` code linked in #177 creates a composited cover only for images which deviate greatly from the 2:3 aspect ratio. The rest (falling within a certain threshold of deviation from 2:3) are stretched using the code I modified, and images which are close enough to 2:3 to be considered "stretchable" are the only ones I would consider to be "croppable". If the goal is to replace stretching where it occurs with cropping, then `resize_cover` (now renamed in https://github.com/kra-mo/cartridges/commit/c2d671273ae47f5dd51cc25e33bdb56817e681cf) would be the natural place to do it, given that this is where the stretching is done.

In any case, I respect the decision not to automate any potentially destructive cropping, as it's easy to imagine cases in which even slight cropping would ruin an image (such as one with a thin border at the edges).

Sorry about the unsolicited pull request, by the way. It's not as if I expected it to merge as-is; I just wanted to demonstrate what I had in mind for #177, and I figured it would probably be rewritten by someone better at coding if the basic idea were accepted.
